Fix two skin-related crashes

garbear opened the pull request:
This PR contains two fixes:

* Fix segfault in guilib when `<itemlayout>` or `<focuslayout>` are missing or incomplete
* Fix segfault on skin reload due to accessing items after their owner has been destroyed

## How Has This Been Tested?
Fixes have been in my branch through most of the game OSD development.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

tamland commented on xbmc/listproviders/StaticProvider.cpp:
```diff
@@ -117,6 +117,6 @@ bool CStaticListProvider::AlwaysFocusDefaultItem() const
 bool CStaticListProvider::OnClick(const CGUIListItemPtr &item)
 {
-  CGUIStaticItemPtr staticItem = std::static_pointer_cast<CGUIStaticItem>(item);
+  CGUIStaticItem *staticItem = static_cast<CGUIStaticItem*>(item.get());
```
Something missing in this commit? This changes nothing

garbear replied:
The problem is `item`'s lifetime ends with the final call in this function. Acquiring a shared pointer means that item is used after `item`'s lifetime ends.

notspiff commented:
Holding a shared ptr extends the item's lifetime. Now you make it dangling
instead which does not sound like an improvement.

10. des. 2017 17:54 skrev "Garrett Brown" <notifications@github.com>:

> *@garbear* commented on this pull request.
> ------------------------------
>
> In xbmc/listproviders/StaticProvider.cpp
> <https://github.com/xbmc/xbmc/pull/13151#discussion_r155953439>:
>
> > @@ -117,6 +117,6 @@ bool CStaticListProvider::AlwaysFocusDefaultItem() const
>
>  bool CStaticListProvider::OnClick(const CGUIListItemPtr &item)
>  {
> -  CGUIStaticItemPtr staticItem = std::static_pointer_cast<CGUIStaticItem>(item);
> +  CGUIStaticItem *staticItem = static_cast<CGUIStaticItem*>(item.get());
>
> The problem is item's lifetime ends with the final call in this function.
> Acquiring a shared pointer means that item is used after item's lifetime
> ends.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/13151#discussion_r155953439>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFktYIvnBaiJuCA5bcirhwpgSz3D__V0ks5s_AyvgaJpZM4Q7f31>
> .
>

garbear commented:
Dangling and no crash is better than dangling and crash

tamland commented:
It's absolutely not. All this does is make it do something undefined instead the correct thing. A shared ptr doesn't dangling nor does any list control 'own' them (they are shared).

garbear commented:
The behavior is only undefined if the object is used after being destroyed. Otherwise `delete this;` would be UB, yet our codebase has 15 instances of that.

tamland commented:
Then what does casting a shared pointer, that's already in scope, to a raw pointer do?

xhaggi reviewed (approved):
Looks good

garbear commented:
> Then what does casting a shared pointer, that's already in scope, to a raw pointer do?

Avoids a copy. The problem with a copy is that, due to RAII, the object's destructor is called when the ref count drops to zero. If we increase the ref count before the object should be destroyed, we call a destructor on a dangling object.

tamland commented:
Then it's UB. If the share pointer is not in scope, you can not use it after the `item.get()`

> The behavior is only undefined if the object is used after being destroyed

Also not true. All un-synchronized access is undefined. Like deleting it while it's being used.

garbear commented:
I'm confused. If that's true, then how is `delete this;` not UB?

tamland commented:
It is UB, if it's not synchronized (and probably many other cases)

garbear commented:
Gotchya. What about this? https://github.com/garbear/xbmc/commit/2ac65f8

I'm just eyeballing it here. I'll try applying soon this and see what happens.

garbear commented:
No luck. I think I've solved a bug like this before. Got some history to look through...

garbear committed:
Fix crash if empty <itemlayout> or <focuslayout> is given

garbear committed:
Fix crash on skin reload

The crash happens when a dialog is opened from a list control. The list
controls own the items, so it is invalid to access them after the list is
destroyed.

MartijnKaijser commented:
jenkins build and merge

garbear commented:
Merging to fix https://trac.kodi.tv/ticket/17935 before release. This is an improvement regardless of how much refactoring should be done.
